File: design-pattern-java/app/src/test/java/design/pattern/java/patterns/structural/composite/CompositeTest.java
package design.pattern.java.patterns.structural.composite;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class CompositeTest {
    class ComponentMock implements UIComponent {
        private Boolean wasCalled;
        public ComponentMock() {
            this.wasCalled = false;
        }
        @Override
        public String render() {
            this.wasCalled = true;
            return "";
        }
        public Boolean hasBeenCalled() {
            return this.wasCalled;
        }

    }
    @Test
    public void shouldCallComponentRenderWhenContainerRender() {
        ComponentMock child = new ComponentMock();
        new ContainerComposite(child).render();
        Assertions.assertTrue(child.hasBeenCalled());
    }
    @Test
    public void shouldContainChildTextMsgWhenRenderContainerIsCalled() {
        TextComponent child = new TextComponent("test");
        String ui = new ContainerComposite(child).render();
        Assertions.assertTrue(ui.contains(child.getMessage()));
    }
}

File: design-pattern-jvm/app/src/test/kotlin/design/pattern/kotlin/patterns/structural/composite/CompositeTest.kt
package design.pattern.kotlin.patterns.structural.composite

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class CompositeTestKt {
    internal inner class ComponentMock : UIComponent {
        private var wasCalled = false
        override fun render(): String {
            wasCalled = true
            return ""
        }

        fun hasBeenCalled(): Boolean {
            return wasCalled
        }
    }

    @Test
    fun shouldCallComponentRenderWhenContainerRender() {
        val child = ComponentMock()
        ContainerComposite(child).render()
        Assertions.assertTrue(child.hasBeenCalled())
    }

    @Test
    fun shouldContainChildTextMsgWhenRenderContainerIsCalled() {
        val child = TextComponent("test")
        val ui = ContainerComposite(child).render()
        Assertions.assertTrue(ui.contains(child.message))
    }
}
